616/Architecture Developer Docs (#147)
microsoft/carbon-aware-sdk: docs/architecture/overview.md

# Architecture Overview

> The Carbon Aware SDK was built in 3 layers, with each layer having it's own set of responsibilities. 

Reviewer: ```suggestion
The Carbon Aware SDK is built with three layers where each layer has its own set of responsibilities. 
```

> 1. User-interface layer

Reviewer 2: NIT: spelling

```suggestion
1. User interface layer
```

> 1. User interface layer

Reviewer: apologies if i'm opening a can of worms here, but any reason we can't call it the "consumer layer"?

Author: resolved by team vote

2. Business Logic Layer
3. Data Layer

Each layer only has knowledge of its own responsibilities, and specific channels of communication between adjacent layers. This allows the SDK to switch to different implementations withing the same layer without any necessary code changes.

> ![Layers architecture diagram](../images/layers-architecture.png)

Reviewer 2: Maintaining this image is going to be painful in the long-term.  Consider including a source `.drawio` file for this diagram.   (The best would be if the `.drawio.png` file renders in markdown and is editable by maintainers.  We'd have to test that)

Reviewer 3: Good call @bderusha . @JenMadiedo , the source file is available too.

Author: @juzuluag is the .drawio.png the source? So if someone remove the .png extension and uploads the .drawio file it could be edited?

Author: resolved offline, adding .drawio.png file

## User Interface Layer

> The user interface layer is the entry point for interacting with the SDK. It handles taking the user input and deciding which aggregator(s) should handle processing the response. There are currently 2 different interfaces that can be used in the SDK

Reviewer 2: ```suggestion
The user interface layer is the entry point for interacting with the SDK. It handles taking the user input and deciding which aggregator(s) should handle processing the input. There are currently 2 different UIs that can be used in the SDK
```

1. A RESTful WebAPI that can be called using HTTP requests

> 2. A CLI that takes in paths and arguments.

Reviewer 2: ```suggestion
2. A command line tool that runs directly on a host machine.
```

> Both interfaces access the same logic in the layers below.

Reviewer 2: ```suggestion
Both UIs access the same logic in the layers below.
```

## Business Logic Layer

> The business logic layer functions as the processor, taking in the user request and figuring out how to fulfill it. It is comprised of a set of aggregators that know what type of data they need and how to calculate the result. 

Reviewer 2: ```suggestion
The business logic layer functions as the processor, taking in the user input and figuring out how to fulfill it. It is comprised of a set of aggregators that know what type of data they need and how to calculate the result. 
```

### Aggregator

> Aggregators have knowledge of the different data source interfaces in the data layer. They can aggregate multiple sources of data into a single response given a request, like the SCI Score aggregator shown in the diagram above. They can also map to a single data source interface, handling all the operations needed by the user for that area of data. 

Reviewer 2: As the SCI score aggregator is not actually implemented, we should make that clear in the diagram and how we talk about it.

> For example, the carbon aware aggregator knows how to handle any user requests for carbon intensity information. For carbon intensity requests, it knows how to calculate the average carbon intensity over a time period, or the best carbon intensity given a set of locations. For forecasted carbon intensity, it knows the difference between a current vs. historical forecast and how what rate to query the data sources.

Reviewer 2: ```suggestion
For example, the `CarbonAwareAggregator` handles requests for various carbon emissions information. It can calculate the average carbon emissions over a time period, or the best carbon emissions given a set of locations. It can transform forecasted carbon emissions to suit particular use-cases. It can also just deliver the emissions data points in a standard schema without performing any calculations.
```

Author: Much clearer, thank you!

## Data Layer
The data layer is responsible for ingesting data into the SDK. It is comprised of a set of data source interfaces representing the broad categories of data used by the SDK. 

### Data Source Interface

> A data source interface defines all the necessary parameters and functions needed to access that category of data. These definitions must be independent of specific data sources and use abstracted input/output parameters expected from the business logic layer. This abstraction enables multiple data sources to extend a single interface and be switched out on the fly based on the user request without code changes.

Reviewer 2: ```suggestion
A data source interface defines all the necessary parameters and functions needed to access that category of data. These definitions must be independent of specific data sources and use abstracted input/output parameters expected from the business logic layer. This abstraction enables multiple data sources to extend a single interface and be switched out and configured based on the needs of the operator.
```

# Architecture Overview

> The Carbon Aware SDK is built with three layers where each layer has its own set of responsibilities.

Reviewer 2: While we are updating the name of the top level of our architecture anyway, we should consider using the more common term, `tier` over `layer`.  This codebase is a classic `N-tier architecture` and will be more easily identified as such if we use that common language.

1. User interface layer
2. Business Logic Layer
3. Data Layer

Each layer only has knowledge of its own responsibilities, and specific channels of communication between adjacent layers. This allows the SDK to switch to different implementations withing the same layer without any necessary code changes.

> ![Layers architecture diagram](../images/layers-architecture.drawio.png)

Reviewer 2: Love this!  I was able to pull it into draw.io with no issues for editing needs.

Reviewer 3: @JenMadiedo , were you able to edit the file? So it can be updated with the new name for the upper layer?

Author: Yes I can edit and will update with push

# Architecture Overview

The Carbon Aware SDK is built with three tiers where each tier has its own set of responsibilities.
1. Consumer Tier
2. Business Logic Tier
3. Data Tier

Each tier only has knowledge of its own responsibilities, and specific channels of communication between adjacent tiers. This allows the SDK to switch to different implementations withing the same tier without any necessary code changes.

![Tiers architecture diagram](../images/tiers-architecture.drawio.png)

## Consumer Tier
The consumer tier is the entry point for interacting with the SDK. It handles taking the user input and deciding which aggregator(s) should handle processing the input. There are currently 2 different consumers that can be used in the SDK
1. A RESTful WebAPI that can be called using HTTP requests.
2. A command line tool that runs directly on a host machine.

> Both consumers access the same logic in the tiers below.

Reviewer 2: ```suggestion
Both consumers access the same components in the tiers below.
```
